File: 编程/leetcode/1464-[E]-数组中两元素的最大乘积/solution.py
from typing import List
import sys


class Solution1:
    def maxProduct(self, nums: List[int]) -> int:
        nums = [x-1 for x in nums]
        nums.sort()
        if nums[-2] > 0:
            return nums[-1]*nums[-2]
        return nums[0]*nums[1]


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_1, max_2 = -1, -1
        for n in nums:
            if n >= max_1:
                max_2 = max_1
                max_1 = n
            elif n >= max_2:
                max_2 = n
            else:
                continue
        return (max_1-1)*(max_2-1)

if __name__ == '__main__':
    test = Solution()
    print(test.maxProduct([3,4,5,2]))
    print(test.maxProduct([1,5,4,5]))
    print(test.maxProduct([3,7]))
    print(test.maxProduct([1,1,1,1]))
    print(test.maxProduct([10,2,5,2]))

File: 编程/leetcode/1464-[E]-数组中两元素的最大乘积/solution.rs
struct Solution;

impl Solution {
    pub fn max_product(nums: Vec<i32>) -> i32 {
        let (mut max_first, mut max_second) = (-1, -1);
        for &n in nums.iter() {
            if n >= max_first {
                max_second = max_first;
                max_first = n;
            } else if n > max_second {
                max_second = n;
            }
        };
        (max_first-1)*(max_second-1)
    }
}


fn main() {
    let result = Solution::max_product([3,4,5,2].to_vec());
    println!("{}", result)
}